### PR TITLE
Properly add order to scripted fields

### PR DESCRIPTION
**Pull request by galkk**
*Issue #, if available:* [104](https://github.com/opendistro-for-elasticsearch/sql/issues/104)

*Description of changes:*

In previous PRs, the fields were associated with corresponding SQLExpressions. During aggregation construction now it's possible to see if group by fields and order by fields are related to the same expression and correctly add order by clause into aggregation query.

For aggregation queries the script fields won't be added into "sort" part of aggregation query

```
POST _opendistro/_sql 
{
  "query" : "SELECT date_format(insert_time, 'dd-MM-YYYY'), min(all_client) FROM online group by date_format(insert_time, 'dd-MM-YYYY') order by date_format(insert_time, 'dd-MM-YYYY') desc"
}
```

generates correct order by in aggregation, and removes that field from order, like it was done previously

```json
  "aggregations": {
    "date_format_1260712542": {
      "terms": {
        "script": {
          "source": "def date_format_1122616994 = DateTimeFormatter.ofPattern('dd-MM-YYYY').withZone(ZoneId.systemDefault()).format(Instant.ofEpochMilli(doc['utc_time'].value.getMillis()));return date_format_1122616994;",
          "lang": "painless"
        },
        "size": 10,
        "min_doc_count": 1,
        "shard_min_doc_count": 0,
        "show_term_doc_count_error": false,
        "order": {
          "_key": "desc"
        }
      },
      "aggregations": {
        **"AVG(bytes)": {
          "avg": {
            "field": "bytes"
          }
        }**
      }
    }
```

Testing: unit test, integration test, manual test via kibana UI

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/query/AggregationQueryAction.java [hunk 1]
```diff
@@ -231,10 +238,34 @@ private AggregationBuilder getGroupAgg(Field field, Select select2) throws SqlPa
...
+        }
+
         return lastAgg;
     }
 
+    private void addOrderByScriptFieldIfPresent(ScriptMethodField field, Select select, AggregationBuilder lastAgg) {
+        // This is a hacky way, but it's the best that could be done now.
+
+        // TODO: Explore other ways to correlate different fields/functions in the query (params?)
+        ScriptMethodField groupByScriptField = field;
```
Comment (dai-chen) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/query/AggregationQueryAction.java]: maybe inline or rename `field` to `groupByScriptField` directly?

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/query/AggregationQueryAction.java [hunk 2]
```diff
@@ -231,10 +238,34 @@ private AggregationBuilder getGroupAgg(Field field, Select select2) throws SqlPa
...
+        SQLExpr groupByExpression = groupByScriptField.getExpression();
+        for (Order order : select.getOrderBys()) {
+            if (order.isScript()) {
+                ScriptMethodField orderByScriptField = (ScriptMethodField) order.getSortField();
+                SQLExpr orderByExpression = orderByScriptField.getExpression();
+
+                if (groupByExpression.equals(orderByExpression)) {
```
Comment (dai-chen) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/query/AggregationQueryAction.java]: I forgot the intent of recent added `expression` in `Field` and have the following questions:

 1. What's information missing in `ScriptMethodField` to force you keep the reference to the expr in original AST?
 2. Does the equality of expression imply the equality of `ScriptMethodField`?

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/query/AggregationQueryAction.java [hunk 3]
```diff
@@ -231,10 +238,34 @@ private AggregationBuilder getGroupAgg(Field field, Select select2) throws SqlPa
...
+
+        // find if we have order for that aggregation. As of now only special case for script fields
+        if (field.isScriptField()) {
+            addOrderByScriptFieldIfPresent((ScriptMethodField) field, select, lastAgg);
+        }
+
         return lastAgg;
     }
 
+    private void addOrderByScriptFieldIfPresent(ScriptMethodField field, Select select, AggregationBuilder lastAgg) {
```
Comment (dai-chen) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/query/AggregationQueryAction.java]: I think it's more consistent with your method name to "narrow down" the second param and pass in `select.getOrderBys()`.

File: src/test/java/com/amazon/opendistroforelasticsearch/sql/esintgtest/DateFormatIT.java [hunk 4]
```diff
@@ -135,6 +137,29 @@ public void sortByDateFormat() throws IOException {
...
+                        "ORDER BY date_format(insert_time, 'dd-MM-YYYY') DESC")
+                .getJSONObject("aggregations");
+
+        String date = DateFormatTest.getScriptAggregationKey(aggregations, "date_format");
+        JSONArray buckets = aggregations.getJSONObject(date).getJSONArray("buckets");
+
+        assertThat(buckets.length(), is(8));
+
+        String previousKey = buckets.getJSONObject(0).getString("key");
+        for (int i=1; i < buckets.length(); i++) {
```
Comment (dai-chen) [inline src/test/java/com/amazon/opendistroforelasticsearch/sql/esintgtest/DateFormatIT.java]: nit: space between `=`. And just a proposal, a custom Hamcrest `Matcher` to check collection is sorted could provide higher abstraction and be helpful to many test cases regarding `ORDER BY`.

File: src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/DateFormatTest.java [hunk 5]
```diff
@@ -88,15 +91,44 @@ public void orderByTest() {
...
+        JSONObject elasticQuery = new JSONObject(elasticDsl);
+
+        JSONObject aggregations = elasticQuery.getJSONObject("aggregations");
+        String dateFormatAggregationKey = getScriptAggregationKey(aggregations, "date_format");
+
+        return aggregations.getJSONObject(dateFormatAggregationKey).getJSONObject("terms").getJSONObject("order");
+    }
+
+    public static String getScriptAggregationKey(JSONObject aggregation, String prefix) {
+        return aggregation.keySet().stream().filter(x -> x.startsWith(prefix)).findFirst().get();
```
Comment (dai-chen) [inline src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/DateFormatTest.java]: Maybe better to use `orElseThrow(Supplier)` to provide more clear error message than NPE if expected value not present.

Comment (galkk) [reply src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/DateFormatTest.java]: Done

File: src/test/java/com/amazon/opendistroforelasticsearch/sql/esintgtest/DateFormatIT.java (hunk 4 above)
Comment (galkk) [reply src/test/java/com/amazon/opendistroforelasticsearch/sql/esintgtest/DateFormatIT.java]: Fixed formatting.
Was able to have oneliner without the need to introduce custom matcher

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/query/AggregationQueryAction.java (hunk 2 above)
Comment (galkk) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/query/AggregationQueryAction.java]: 1. Script method field had only generated script information, and due to generated field names inside of the script the code can't understand that those generated scripts are actually related to the same expression.
2. No, script method field has more information, for example field alias

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/query/AggregationQueryAction.java (hunk 3 above)
Comment (galkk) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/query/AggregationQueryAction.java]: Done

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/query/AggregationQueryAction.java (hunk 2 above)
Comment (dai-chen) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/query/AggregationQueryAction.java]: Thanks for clarification.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/query/AggregationQueryAction.java [hunk 6]
```diff
@@ -231,10 +239,34 @@ private AggregationBuilder getGroupAgg(Field field, Select select2) throws SqlPa
...
+                    .getOrderBys()
+                    .stream()
+                    .filter(order -> groupByExpression.equals(order.getSortField().getExpression()))
+                    .findFirst();
+
+        if (!orderForGroupBy.isPresent()) {
+            return;
+        }
+
+        TermsAggregationBuilder termsAggregationBuilder = (TermsAggregationBuilder) groupByAggregation;
```
Comment (dai-chen) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/query/AggregationQueryAction.java]: Just an option, you may be able to put this logic into  `ifPresent(Consumer)` which could append after `.findFirst()` in chain. And the `if` check above could be removed. Then the code would be exactly consistent with your method name.

```
select...
      .findFirst() 
      .ifPresent(...);
```

File: src/test/java/com/amazon/opendistroforelasticsearch/sql/esintgtest/DateFormatIT.java [hunk 7]
```diff
@@ -135,6 +140,28 @@ public void sortByDateFormat() throws IOException {
...
+        JSONArray buckets = aggregations.getJSONObject(date).getJSONArray("buckets");
+
+        assertThat(buckets.length(), is(8));
+
+        List<String> aggregationSortKeys = IntStream.range(0, 8)
+                .mapToObj(index -> buckets.getJSONObject(index).getString("key"))
+                .collect(Collectors.toList());
+
+        assertTrue("The query result must be sorted by date in descending order",
+                Ordering.natural().reverse().isOrdered(aggregationSortKeys));
```
Comment (dai-chen) [inline src/test/java/com/amazon/opendistroforelasticsearch/sql/esintgtest/DateFormatIT.java]: Good to know this!

File: src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/DateFormatTest.java [hunk 8]
```diff
@@ -88,15 +91,47 @@ public void orderByTest() {
...
+        String elasticDsl = queryAction.explain().explain();
+        JSONObject elasticQuery = new JSONObject(elasticDsl);
+
+        JSONObject aggregations = elasticQuery.getJSONObject("aggregations");
+        String dateFormatAggregationKey = getScriptAggregationKey(aggregations, "date_format");
+
+        return aggregations.getJSONObject(dateFormatAggregationKey).getJSONObject("terms").getJSONObject("order");
+    }
+
+    public static String getScriptAggregationKey(JSONObject aggregation, String prefix) {
```
Comment (dai-chen) [inline src/test/java/com/amazon/opendistroforelasticsearch/sql/unittest/DateFormatTest.java]: nit: need a little more reformat in this method (don't know how to add multiple-line comment here...)

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/query/AggregationQueryAction.java (hunk 6 above)
Comment (galkk) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/query/AggregationQueryAction.java]: Good idea, thanks. Done

**Pull request merged**